iree/compiler/Dialect/Flow/Transforms/DispatchLinalgOnTensors.cpp

```diff
@@ -0,0 +1,309 @@
+                                 clonedRoot->getNumResults()))) {
+      b.create<IREE::Flow::DispatchOutputStoreOp>(loc, std::get<0>(it),
+                                                  std::get<1>(it), llvm::None,
+                                                  llvm::None, llvm::None);
+    }
+    // TODO(nicolasvasilache): return `clonedRoot->getResults()` once we have
+    // shape operands and we drop tie_shape.
+    b.create<IREE::Flow::ReturnOp>(loc);
+  }
+  return dispatchOp;
+}
+
+// Rewrite pattern to ensure only ops with tensor semantics are tiled.
+struct TileAndDistributeOnTensorsPattern
+    : public linalg::LinalgBaseTilingPattern {
+  using Base = linalg::LinalgBaseTilingPattern;
+  TileAndDistributeOnTensorsPattern(linalg::LinalgTilingOptions options,
+                                    linalg::LinalgMarker marker,
+                                    PatternBenefit benefit = 1)
+      : Base(options, marker, benefit) {}
+
+  LogicalResult matchAndRewrite(Operation *op,
+                                PatternRewriter &rewriter) const override {
+    auto linalgOp = dyn_cast<linalg::LinalgOp>(op);
+    if (!linalgOp || !linalgOp.hasTensorSemantics()) return failure();
+
+    linalg::LinalgOp clonedLinalgOp;
+    Operation *dispatch =
+        buildFlowWorkgroupDispatchOp(rewriter, linalgOp, clonedLinalgOp);
+
+    OpBuilder::InsertionGuard g(rewriter);
+    rewriter.setInsertionPoint(clonedLinalgOp);
+    SmallVector<Value, 4> tensorResults;
+    if (failed(Base::matchAndRewriteBase(clonedLinalgOp, rewriter,
+                                         tensorResults))) {
+      // Need to erase on failure to avoid infinite loop
+      rewriter.eraseOp(dispatch);
+      return failure();
+    }
+    rewriter.replaceOp(op, dispatch->getResults());
+    rewriter.replaceOp(clonedLinalgOp, tensorResults);
+
+    return success();
+  }
+};
+
+static Value buildFlowWorkgroupIdOp(OpBuilder &b, unsigned dim) {
+  return b.create<IREE::Flow::DispatchWorkgroupIDOp>(
+      b.getInsertionPoint()->getLoc(), dim);
+}
+
+static Value buildFlowWorkgroupCountOp(OpBuilder &b, unsigned dim) {
+  return b.create<IREE::Flow::DispatchWorkgroupCountOp>(
+      b.getInsertionPoint()->getLoc(), dim);
+}
+
+// After outlining in dispatch region we can rewrite the dispatch ops with
+// proper captures.
+static void setDispatchWorkgroupOperands(
+    IREE::Flow::DispatchWorkgroupsOp dispatchOp) {
+  llvm::SetVector<Value> valuesSet;
+  mlir::getUsedValuesDefinedAbove(dispatchOp.body(), valuesSet);
+  ValueRange valuesDefinedAbove{valuesSet.getArrayRef()};
+
+  SmallVector<Value, 4> originalDispatchOperands = dispatchOp.operands();
+  llvm::SetVector<Value> originalDispatchOperandsSet(
+      originalDispatchOperands.begin(), originalDispatchOperands.end());
+  SmallVector<Value, 4> newOperands, clones;
+  for (Value v : valuesDefinedAbove) {
+    if (v.getDefiningOp<ConstantOp>()) {
```

Comment:
I dont think we do this for all constants. Only for splat constants.

Reply:
Good point, added a level of indirection and only supporting ConstantIndex/Int/Float for now.
We can extend on a per-need basis.